url0307c0e995_js_labs_js_collider_tgz-pJ8-org_jsl_tests_shmem_throughput_MainJ8/src/org/jsl/collider/ColliderImpl.java

```diff
@@ -0,0 +1,798 @@
+                 *     but will present in the keys set till the next
+                 *     Selector.select() call.
+                 * }
+                 */
+            }
+            m_run = false;
+            return 0;
+        }
+    }
+
+    private static class DummyRunnable extends SelectorThreadRunnable
+    {
+        public int runInSelectorThread()
+        {
+            return 0;
+        }
+    }
+
+    private class SelectorAlarm extends ThreadPool.Runnable
+    {
+        public SelectorThreadRunnable cmp;
+
+        public SelectorAlarm( SelectorThreadRunnable runnable )
+        {
+            cmp = runnable;
+        }
+
+        public void runInThreadPool()
+        {
+            if (m_strHead == cmp)
+                m_selector.wakeup();
+            cmp = null;
+            m_alarm.compareAndSet( null, this );
+        }
+    }
+
+    private static class SessionSharedData
+    {
+        private final RetainableDataBlockCache m_inputQueueDataBlockCache;
+        private final int m_joinMessageMaxSize;
+        private final RetainableByteBufferPool m_joinPool;
+
+        public SessionSharedData(
+                RetainableDataBlockCache inputQueueDataBlockCache,
+                int joinMessageMaxSize,
+                RetainableByteBufferPool joinPool )
+        {
+            m_inputQueueDataBlockCache = inputQueueDataBlockCache;
+            m_joinMessageMaxSize = joinMessageMaxSize;
+            m_joinPool = joinPool;
+        }
+
+        RetainableDataBlockCache getInputQueueDataBlockCache() { return m_inputQueueDataBlockCache; }
+        int getJoinMessageMaxSize() { return m_joinMessageMaxSize; }
+        RetainableByteBufferPool getJoinPool() { return m_joinPool; }
+    }
+
+    private SessionSharedData getSessionSharedData( final SessionEmitter sessionEmitter )
+    {
+        final Config config = getConfig();
+
+        int inputQueueBlockSize = sessionEmitter.inputQueueBlockSize;
+        if (inputQueueBlockSize == 0)
+        {
+            inputQueueBlockSize = config.inputQueueBlockSize;
+            assert( inputQueueBlockSize > 0 );
+        }
+
+        int joinMessageMaxSize = sessionEmitter.joinMessageMaxSize;
+        if (joinMessageMaxSize < 0)
+        {
+            joinMessageMaxSize = config.joinMessageMaxSize;
+            if (joinMessageMaxSize < 0)
+                joinMessageMaxSize = 0;
+        }
+
+        m_lock.lock();
+        try
+        {
+            RetainableDataBlockCache cache = m_dataBlockCache.get( inputQueueBlockSize );
+            if (cache == null)
+            {
+                cache = new RetainableDataBlockCache(
+                            config.useDirectBuffers,
+                            inputQueueBlockSize,
+                            8 /* initial size */,
+                            config.inputQueueCacheMaxSize );
+                m_dataBlockCache.put( inputQueueBlockSize, cache );
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 234:242, 721:729.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.